subsys/logging/log_core.c

```diff
@@ -62,11 +64,117 @@ static k_tid_t proc_tid;
 static u32_t dummy_timestamp(void);
 static timestamp_get_t timestamp_func = dummy_timestamp;
 
+
+bool log_is_strdup(const void *buf);
+
 static u32_t dummy_timestamp(void)
 {
 	return 0;
 }
 
+/**
+ * @brief Count number of string format specifiers (%s).
+ *
+ * Result is stored as the mask (argument n is n'th bit). Bit is set if %s was
+ * found.
+ *
+ * @param str String.
+ * @param nargs Number of arguments in the string.
+ *
+ * @return Mask with %s format specifiers found.
+ */
+static u32_t count_s(const char *str, u32_t nargs)
+{
+	__ASSERT_NO_MSG(nargs <= 8*sizeof(u32_t));
+	char curr = *str++;
+	bool arm = (curr == '%');
+	u32_t arg = 0;
+	u32_t mask = 0;
+
+	while (curr) {
+		curr = *str++;
+		if (arm) {
+			if (curr == 's') {
+				mask |= 1 << arg++;
+			} else if (curr != '%') {
+				arg++;
+			}
+
+			arm = false;
+		} else if (curr == '%') {
+			arm = true;
+		}
+
+		if (arg == nargs) {
+			return mask;
+		}
+	}
+
+	return mask;
+}
+
+/**
+ * @brief Check if address is in read only section.
+ *
+ * @param addr Address.
+ *
+ * @return True if address identified within read only section.
+ */
+static bool is_rodata(const void *addr)
+{
+#if defined(CONFIG_ARM) || defined(CONFIG_ARC) || defined(CONIFG_RISCV32) || \
+	defined(CONFIG_X86)
+	extern const char *_image_rodata_start[];
+	extern const char *_image_rodata_end[];
+	#define RO_START _image_rodata_start
+	#define RO_END _image_rodata_end
+#elif defined(CONFIG_NIOS2)
+	extern const char *_image_rom_start[];
+	extern const char *_image_rom_end[];
+	#define RO_START _image_rom_start
+	#define RO_END _image_rom_end
+#elif defined(CONFIG_XTENSA)
+	extern const char *_rodata_start[];
+	extern const char *_rodata_end[];
+	#define RO_START _rodata_start
+	#define RO_END _rodata_end
+#else
+	#define RO_START 0
+	#define RO_END 0
+#endif
+
+	return ((addr >= (void *)RO_START) && (addr < (void *)RO_END));
```

Comment:
This I believe is wrong as standard does not allow relational operators on void pointers (you can use equality operators though).

6.5.8 in http://www.open-std.org/jtc1/sc22/wg14/www/docs/n1256.pdf says you can compare pointers to specific type and these must be related (check https://stackoverflow.com/questions/24807453/less-than-comparison-for-void-pointers).
I think it works without any problem but I think it works because we have linear memory addressing (no segments) and because of GCC extension which assumes `sizeof(void)` is 1 (https://gcc.gnu.org/onlinedocs/gcc-4.8.0/gcc/Pointer-Arith.html). But this may not always work as it is essentially and undefined behavior in C.
https://www.quora.com/How-do-I-compare-a-void-pointer-in-C-with-another-pointer

I think it should on ARM unless somebody would want to abuse the UB for code optimization. It would be safer however to cast `addr` to `const char *` to make sure pointer arithmetic works.